contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,850 @@
+
+        // get msg hash
+        bytes32 msgHash = keccak256(abi.encodePacked(voteAddress, block.chainid));
+        bytes memory msgBz = new bytes(32);
+        assembly {
+            mstore(add(msgBz, 32), msgHash)
+        }
+
+        // call the precompiled contract to verify the BLS signature
+        // the precompiled contract's address is 0x66
+        bytes memory input = bytes.concat(msgBz, blsProof, voteAddress); // length: 32 + 96 + 48 = 176
+        bytes memory output = new bytes(1);
+        assembly {
+            let len := mload(input)
+            if iszero(staticcall(not(0), 0x66, add(input, 0x20), len, add(output, 0x20), 0x01)) { revert(0, 0) }
+        }
+        uint8 result = uint8(output[0]);
+        if (result != uint8(1)) {
+            return false;
+        }
+        return true;
+    }
+
+    function _deployStakeCredit(address operatorAddress, string memory moniker) internal returns (address) {
+        address creditProxy = address(new TransparentUpgradeableProxy(STAKE_CREDIT_ADDR, DEAD_ADDRESS, ""));
+        IStakeCredit(creditProxy).initialize{ value: msg.value }(operatorAddress, moniker);
+
+        return creditProxy;
+    }
+
+    function _checkValidatorSelfDelegation(address operatorAddress) internal {
+        Validator storage valInfo = _validators[operatorAddress];
+        if (valInfo.jailed) {
+            return;
+        }
+        if (IStakeCredit(valInfo.creditContract).getSelfDelegationBNB() < minSelfDelegationBNB) {
+            _jailValidator(valInfo);
+            // need to inform validator set contract to remove the validator
+            IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).jailValidator(valInfo.consensusAddress);
+        }
+    }
+
+    function _updateSlashRecord(
+        address operatorAddress,
+        uint256 slashAmount,
+        uint256 height,
+        SlashType slashType
+    ) internal returns (SlashRecord storage) {
+        bytes32 slashKey = _getSlashKey(operatorAddress, height, slashType);
+        SlashRecord storage record = _slashRecords[slashKey];
+        require(record.slashHeight == 0, "SLASHED");
+
+        record.jailUntil = block.timestamp + doubleSignJailTime;
+        record.slashAmount = slashAmount;
+        record.slashHeight = uint248(height);
+        record.slashType = slashType;
+
+        return record;
+    }
+
+    function _jailValidator(Validator storage valInfo) internal {
+        // keep the last eligible validator
+        bool isLast = (numOfJailed >= _validatorSet.length() - 1);
+        if (isLast) {
+            emit ValidatorEmptyJailed(valInfo.operatorAddress);
+            return;
+        }
+
+        if (!valInfo.jailed) {
+            valInfo.jailed = true;
+            numOfJailed += 1;
+        }
+        emit ValidatorJailed(valInfo.operatorAddress);
+
+        // remove from the eligibleValidatorSet if exists
+        for (uint256 i; i < _eligibleValidatorSet.length; ++i) {
```

Comment:
I dont think it is necessary to remove the items.

Reply:
If we don't remove it here, the following scenario might occur: 
During the transition period, `StakeHub` updates the validator set after UTC 00:00, while `BSCValidatorSet` hasn't updated yet (due to the cross-chain package not arriving), and at this point, a validator who has been jailed may still be selected for the next epoch.